module/builder/collection/rate_limiter.go

```diff
@@ -0,0 +1,103 @@
+	unlimited map[flow.Address]struct{}
+	// height of the collection we are building
+	height uint64
+
+	// for each payer, height of latest collection in which a transaction for
+	// which they were payer was included
+	latestCollectionHeight map[flow.Address]uint64
+	// number of transactions included in the currently built block per payer
+	txIncludedCount map[flow.Address]uint
+}
+
+func newRateLimiter(conf Config, height uint64) *rateLimiter {
+	limiter := &rateLimiter{
+		rate:                   conf.MaxPayerTransactionRate,
+		unlimited:              conf.UnlimitedPayers,
+		height:                 height,
+		latestCollectionHeight: make(map[flow.Address]uint64),
+		txIncludedCount:        make(map[flow.Address]uint),
+	}
+	return limiter
+}
+
+// note the existence and height of a transaction in an ancestor collection.
+func (limiter *rateLimiter) addAncestor(height uint64, tx *flow.TransactionBody) {
+
+	// skip tracking payers if we aren't rate-limiting or are configured
+	// to allow multiple transactions per payer per collection
+	if limiter.rate >= 1 || limiter.rate <= 0 {
+		return
+	}
+
+	latest := limiter.latestCollectionHeight[tx.Payer]
+	if height >= latest {
+		limiter.latestCollectionHeight[tx.Payer] = height
+	}
+}
+
+// note that we have added a transaction to the collection under construction.
+func (limiter *rateLimiter) transactionIncluded(tx *flow.TransactionBody) {
+	limiter.txIncludedCount[tx.Payer]++
+}
+
+// applies the rate limiting rules, returning whether the transaction should be
+// omitted from the collection under construction.
+func (limiter *rateLimiter) shouldRateLimit(tx *flow.TransactionBody) bool {
+
+	payer := tx.Payer
+
+	// skip rate limiting if it is turned off or the payer is unlimited
+	_, isUnlimited := limiter.unlimited[payer]
+	if limiter.rate == 0 || isUnlimited {
+		return false
+	}
+
+	// if rate >=1, we only consider the current collection and rate limit once
+	// the number of transactions for the payer exceeds rate
+	if limiter.rate >= 1 {
+		if limiter.txIncludedCount[payer] >= uint(math.Floor(limiter.rate)) {
+			return true
+		}
+	}
+
+	// if rate < 1, we need to look back to see when a transaction by this payer
+	// was most recently included - we rate limit if the # of collections since
+	// the payer's last transaction is less than ceil(1/rate)
+	if limiter.rate < 1 {
+
+		// rate limit if we've already include a transaction for this payer, we allow
+		// AT MOST one transaction per payer in a given collection
+		if limiter.txIncludedCount[payer] > 0 {
+			return true
+		}
+
+		// otherwise, check whether sufficiently many empty collection
+		// have been built since the last transaction from the payer
+
+		latestHeight, hasLatest := limiter.latestCollectionHeight[payer]
+		// if there is no recent transaction, don't rate limit
+		if !hasLatest {
+			return false
+		}
+
+		if limiter.height-latestHeight < uint64(math.Ceil(1/limiter.rate)) {
```

Comment:
This looks like a different heuristic than the idea of "rate". I though we would allow one tx per *fixed* window of `ceil[1/k]` collections, but the implemented method allows one tx per *sliding* window, which is a tougher condition on payers. 

For instance if I consider 25% "rates" of `1`s in a slice, I thought this slice should be accepted `[0,0,1,0,1,0,0,0]`, but in the current code it will be rejected. 

Of course there is no security issue here as these are all heuristics, but just wanted to make sure the code is doing what we want.

Reply:
I thought about computing the rate over a larger timespan, but decided against it. 1 reason is that we open ourselves up to essentially the opposite problem of what you mentioned. For example, if we determine the rate over the past 100 blocks, and the limit is 25% then someone could send 0 transactions for 100 blocks, then send 25 all at once, which is behaviour we want to avoid. The other reason is that when a payer exceeds the limit, we will still include their transaction, it will just remain pending for longer. From your example `[0,0,1,0,1,0,0,0]`, it isn't that we would _reject_ this assignment of transactions to blocks, it is that we would not _produce_ this assignment of transactions to blocks, instead we would produce something like `[0,0,1,0,0,0,1,0]`, where the second transaction (1) would stay in the mempool for a bit longer than it otherwise would.

Reply:
You're right, if we compute the average over a larger window we would allow the payer to include many txs in one collection. 
What I meant with fixed window, is to consider the same small window (4 in case of 25%) and allow 1 tx per window. In the example `[0,0,1,0,1,0,0,0]`, we would consider 2 windows, slice[0:4] and slice[4:8] and each can only have one tx. The sliding window method considers the window [2,6] instead.

Thanks for clarifying the mempool process, I didn't understand that before.